src/fairness/writer/data_writer_db.cpp

```diff
@@ -0,0 +1,165 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#include "src/fairness/writer/data_writer_db.hpp"
+
+using namespace Flux::accounting;
+using namespace Flux::writer;
+
+/******************************************************************************
+ *                                                                            *
+ *                         Private DB Writer API                              *
+ *                                                                            *
+ *****************************************************************************/
+
+sqlite3* data_writer_db_t::open_db (const std::string &path)
+{
+    int rc;
+    sqlite3 *DB = nullptr;
+
+    // open flux-accounting DB in read-write mode
+    rc = sqlite3_open_v2 (path.c_str (), &DB, SQLITE_OPEN_READWRITE, NULL);
+    if (rc != SQLITE_OK) {
+        m_err_msg = "error opening DB: " + std::string (sqlite3_errmsg (DB));
+        errno = EIO;
+
+        return nullptr;
+    }
+
+    return DB;
+}
+
+
+sqlite3_stmt* data_writer_db_t::compile_stmt (sqlite3 *DB, const std::string &s)
+{
+    int rc;
+    sqlite3_stmt *c_stmt = nullptr;
+
+    // compile SQL statement
+    rc = sqlite3_prepare_v2 (DB, s.c_str (), -1, &c_stmt, 0);
+    if (rc != SQLITE_OK) {
+        m_err_msg = sqlite3_errmsg (DB);
+        errno = EINVAL;
+
+        return nullptr;
+    }
+
+    return c_stmt;
+}
+
+
+sqlite3_stmt* data_writer_db_t::bind_param (sqlite3 *DB,
+                                            sqlite3_stmt *c_stmt,
+                                            int index,
+                                            const char *param)
+{
+    int rc;
+
+    // bind parameter to compiled SQL statement
+    rc = sqlite3_bind_text (c_stmt, index, param, -1, NULL);
+    if (rc != SQLITE_OK) {
+        m_err_msg = std::string (sqlite3_errmsg (DB)) + "\n";
+        errno = EINVAL;
+
+        return nullptr;
+    }
+
+    return c_stmt;
+}
+
+
+int data_writer_db_t::update_fairshare_values (
+                                    sqlite3 *DB,
+                                    std::shared_ptr<weighted_tree_node_t> node)
+{
+    int rc;
+    std::string ud;
+    sqlite3_stmt *c_ud = nullptr;
+    const char *fshare, *username, *bank;
+
+    ud = "UPDATE association_table SET fairshare=? WHERE username=? AND bank=?";
+
+    if (node->is_user ()) {
+        // get parameters for UPDATE statement
+        fshare = std::to_string (node->get_fshare ()).c_str ();
+        username = node->get_name ().c_str ();
+        bank = node->get_parent ()->get_name ().c_str ();
+
+        c_ud = compile_stmt (DB, ud);
```

Comment:
Rather than recompiling the sqlite statement on each call, could you compile it once, pass it in as an argument, and call `sqlite3_reset` on it here before re-binding parameters to it?

Reply:
Yes, I can do that! I went ahead and pushed a `fixup` commit 40009bb added a small helper function `clear_and_reset_bindings ()` that calls `sqlite3_reset` before passing the compiled SQL statement back to the recursive function.